cpp/src/parquet/arrow/schema.cc

```diff
@@ -984,21 +992,51 @@ Result<bool> ApplyOriginalMetadata(const Field& origin_field, SchemaField* infer
   bool modified = false;
 
   auto& origin_type = origin_field.type();
+  auto inferred_type = inferred->field->type();
 
   if (origin_type->id() == ::arrow::Type::EXTENSION) {
     const auto& ex_type = checked_cast<const ::arrow::ExtensionType&>(*origin_type);
-    auto origin_storage_field = origin_field.WithType(ex_type.storage_type());
+    if (inferred_type->id() != ::arrow::Type::EXTENSION &&
+        ex_type.extension_name() == ::arrow::extension::JsonExtensionType::type_name()) {
+      // Schema mismatch.
+      //
+      // Arrow extensions are DISABLED in Parquet.
+      // origin_type is ::arrow::extension::json()
+      // inferred_type is ::arrow::binary()
+      //
+      // Origin type is restored as Arrow should be considered the source of truth.
+      DCHECK_EQ(inferred_type->id(), ::arrow::Type::BINARY);
+      inferred->field = inferred->field->WithType(origin_type);
+      RETURN_NOT_OK(ApplyOriginalStorageMetadata(origin_field, inferred));
+    } else {
+      auto origin_storage_field = origin_field.WithType(ex_type.storage_type());
 
-    // Apply metadata recursively to storage type
-    RETURN_NOT_OK(ApplyOriginalStorageMetadata(*origin_storage_field, inferred));
+      // Apply metadata recursively to storage type
+      RETURN_NOT_OK(ApplyOriginalStorageMetadata(*origin_storage_field, inferred));
 
-    // Restore extension type, if the storage type is the same as inferred
-    // from the Parquet type
-    if (ex_type.storage_type()->Equals(*inferred->field->type())) {
-      inferred->field = inferred->field->WithType(origin_type);
+      // Restore extension type, if the storage type is the same as inferred
+      // from the Parquet type
+      if (ex_type.storage_type()->Equals(*inferred->field->type())) {
+        inferred->field = inferred->field->WithType(origin_type);
+      }
     }
     modified = true;
   } else {
+    if (inferred_type->id() == ::arrow::Type::EXTENSION) {
+      const auto& ex_type = checked_cast<const ::arrow::ExtensionType&>(*inferred_type);
+      if (ex_type.extension_name() ==
+          ::arrow::extension::JsonExtensionType::type_name()) {
+        // Schema mismatch.
+        //
+        // Arrow extensions are ENABLED in Parquet.
+        // origin_type is ::arrow::binary()
+        // inferred_type is ::arrow::extension::json()
```

Comment:
In which case does this happen?

Reply:
If the system that wrote the Parquet file did not have JSONExtensionType but had a user created ExtensionType with type name "arrow.json"?

Reply:
Hm or perhaps not. Not sure, I'm ok to remove this.